src/main/java/com/minecolonies/coremod/colony/ColonyList.java

```diff
@@ -0,0 +1,256 @@
+package com.minecolonies.coremod.colony;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.minecolonies.coremod.util.Log;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+//TODO extend list/collection
+/**
+ * Data structure for storing colonies, optimized for performance.
+ *
+ * @param <T> Type of IColony (Colony or ColonyView)
+ *
+ * @author Colton
+ */
+public final class ColonyList<T extends IColony> implements Iterable<T>
+{
+    @VisibleForTesting
+    static final int INITIAL_SIZE = 16;
+
+    private IColony[] list = new IColony[INITIAL_SIZE];
+
+    private final List<Integer> nullIndices = new ArrayList<>();
+
+    private int topID = 0;
+
+    private int size = 0;
+
+    /**
+     * Get the Colony with the provided colony id.
+     * @param index colony id.
+     * @return The Colony associated with the provided id.
+     */
+    @Nullable
+    // no way to remove this, java does it too
+    @SuppressWarnings("unchecked")
+    public T get(int index)
+    {
+        if (index < 1 || index > list.length)
+        {
+            return null;
+        }
+
+        return (T) list[index];
+    }
+
+    /**
+     * Creates a new Colony, adds it to the list, and returns it.
+     * @param world The world for the Colony.
+     * @param position The position for the Colony center.
+     * @return The newly created Colony.
+     */
+    @NotNull
+    public Colony create(World world, BlockPos position)
+    {
+        int colonyID = getNextColonyID();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'colonyID' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)